File: 백준/Gold/2320. 끝말잇기/끝말잇기.py
import sys
sys.setrecursionlimit(1<<20)

def dp(state, x):
    if (state, x) in memo:
        return memo[(state, x)]
    total = lens[x]
    for nx in edges[x]:
        if not state&(1<<nx):
            new_state = state | (1<<nx)
            total = max(total, lens[x] + dp(new_state, nx))
    memo[(state, x)] = total
    return total

n = int(sys.stdin.readline())
words = []
lens = []
edges = [[] for _ in range(n)]
memo = {}

for _ in range(n):
    word = sys.stdin.readline().rstrip()
    words.append((word[0], word[-1]))
    lens.append(len(word))

for i in range(n):
    for j in range(i+1, n):
        if words[i][0] == words[j][1]:
            edges[j].append(i)
        if words[i][1] == words[j][0]:
            edges[i].append(j)
words.clear()

ans = 0
for i in range(n):
    ans = max(ans, dp(1<<i, i))

print(ans)

File: 백준/Gold/2320. 끝말잇기/끝말잇기.c
# include <stdio.h>
# include <string.h>
# define MAX 1<<16

int init(int);
int dfs(int, int);
int n, dp[MAX][16];
typedef struct{
    int len;
    char start, end;
} Word;
Word words[MAX];

int main(void){
    scanf("%d", &n);
    init(n);
    int i, ans = 0;
    for (i=0; i<n; i++){
        int cmp = dfs(1<<i, i);
        if (cmp > ans)
            ans = cmp;
    }
    printf("%d\n", ans);
}

int init(int n){
    for (int i=0; i<n; i++){
        char word[105];
        scanf("%s", word);
        words[i].len = strlen(word);
        words[i].start = word[0];
        words[i].end = word[words[i].len-1];
    }
    for (int i=0; i<1<<n; i++){
        for (int j=0; j<n; j++)
            dp[i][j] = 0;
    }
    return 0;
}

int dfs(int mask, int x){
    if (dp[mask][x])
        return dp[mask][x];
    int total = words[x].len;
    for (int nx=0; nx<n; nx++){
        if (nx==n)
            continue;
        if (!(mask&(1<<nx)) && words[x].end == words[nx].start){
            int new_mask = mask | (1<<nx);
            int cmp = words[x].len + dfs(new_mask, nx);
            if (cmp>total)
                total = cmp;
        }
    }
    dp[mask][x] = total;
    return total;
}